metricplugin/http.go

```diff
@@ -0,0 +1,326 @@
+	ListenAddresses []string `json:"ListenAddresses"`
+}
+
+// API path constants.
+const (
+	APIBasePath                = "/metric_plugin/v1"
+	APIPingPath                = APIBasePath + "/ping"
+	APIMonitoringAddressesPath = APIBasePath + "/monitoring_addresses"
+)
+
+type httpServer struct {
+	api RPCAPI
+
+	// A mutex to coordinate shutting down.
+	closingLock sync.Mutex
+	// A channel shared with all clients to notify them if we're shutting down.
+	closing chan struct{}
+	// A wait group to keep track of the clients.
+	wg sync.WaitGroup
+
+	// The underlying listeners.
+	listeners []net.Listener
+}
+
+func newHTTPServer(api RPCAPI, cfg HTTPServerConfig) (*httpServer, error) {
+	if len(cfg.ListenAddresses) == 0 {
+		return nil, errors.New("missing ListenAddresses")
+	}
+
+	var listeners []net.Listener
+	for _, addr := range cfg.ListenAddresses {
+		listener, err := net.Listen("tcp", addr)
+		if err != nil {
+			return nil, errors.Wrap(err, "unable to bind")
+		}
+		log.Infof("metric export HTTP API running on http://%s%s", listener.Addr(), APIBasePath)
+
+		listeners = append(listeners, listener)
+	}
+
+	return &httpServer{
+		api:       api,
+		listeners: listeners,
+		closing:   make(chan struct{}),
+		wg:        sync.WaitGroup{},
+	}, nil
+}
+
+// Shutdown shuts down the HTTP server.
+// This will close the listeners and wait for currently in-flight requests to
+// finish.
+func (s *httpServer) Shutdown() {
+	s.closingLock.Lock()
+	select {
+	case <-s.closing:
+		// Already closing/closed, no-op.
+		s.closingLock.Unlock()
+		return
+	default:
+		close(s.closing)
+	}
+	s.closingLock.Unlock()
+
+	log.Info("shutting down HTTP server")
+
+	// Shut down listeners.
+	for _, l := range s.listeners {
+		err := l.Close()
+		if err != nil {
+			log.Errorf("error closing listener: %s", err)
+		}
+	}
+
+	// Wait for clients/in-progress calls to shut down.
+	s.wg.Wait()
+	log.Info("HTTP server shut down")
+}
+
+// StartServer starts the HTTP server.
+func (s *httpServer) StartServer() {
+	router := s.buildRoutes()
+
+	for _, l := range s.listeners {
+		s.wg.Add(1)
+		go func(listener net.Listener) {
+			defer s.wg.Done()
+			err := http.Serve(listener, router)
+			if err != nil {
+				log.Errorf("unable to serve: %s", err)
+			}
+		}(l)
+	}
+}
+
+func (s *httpServer) buildRoutes() *httprouter.Router {
+	router := httprouter.New()
+
+	router.GET("/", func(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
+		_, _ = fmt.Fprintf(w, "check out %s\n", APIBasePath)
+	})
+	router.GET(APIBasePath, func(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
+		_, _ = fmt.Fprintf(w, "Methods:\n"+
+			"GET  %s no-op\n"+
+			"GET  %s returns a list of addresses on which the monitor is listening for Bitswap monitoring connections",
+			APIPingPath,
+			APIMonitoringAddressesPath)
+	})
+
+	router.GET(APIPingPath,
+		s.buildHandler(ping(s.api)))
+
+	router.GET(APIMonitoringAddressesPath,
+		s.buildHandler(monitoringAddresses(s.api)))
```

Comment:
I'm wondering if the extra encapsulation between http server and plugin is really necessary in this case, i.e., if the http server really has to access the plugin via the API instead of accessing it's internals.
This feels to me like it makes adding functions to the API more complex, with a lot of boilerplate.
Not suggesting a change here, just for the record

Reply:
Hmm, I agree it adds some boilerplate. I think it's nice to have an the API interface, which easily shows what RPC functionality the plugin provides. And it would be a shame not to use it, then :)